src/runtime_src/core/pcie/tools/xbutil/xbutil.cpp

```diff
@@ -950,33 +954,35 @@ int runShellCmd(const std::string& cmd, std::string& output)
         ret = -EINVAL;
     }
 
-    dup2(stderr_fds[1], 2);
-    std::shared_ptr<FILE> pipe(popen(cmd.c_str(), "r"), pclose);
+    // Save stderr
+    int stderr_save = dup(STDERR_FILENO);
+    // Close existing stderr and set it to be the write end of the pipe.
+    // After fork below, our child process's stderr will point to the same fd.
+    dup2(stderr_fds[1], STDERR_FILENO);
     close(stderr_fds[1]);
+    std::shared_ptr<FILE> stderr_child(fdopen(stderr_fds[0], "r"), fclose);
+    std::shared_ptr<FILE> stdout_child(popen(cmd.c_str(), "r"), pclose);
+    // Restore our normal stderr
+    dup2(stderr_save, STDERR_FILENO);
+    close(stderr_save);
 
-    if (pipe == nullptr) {
+    if (stdout_child == nullptr) {
         std::cout << "ERROR: Failed to run " << cmd << std::endl;
-        close(stderr_fds[0]);
         ret = -EINVAL;
     }
 
-    // Read all output
-    char buf[256];
-    while (ret == 0 && !feof(pipe.get())) {
-        if (fgets(buf, sizeof (buf), pipe.get()) != nullptr) {
+    // Read child's stdout and stderr without parsing the content
+    char buf[1024];
+    while (ret == 0 && !feof(stdout_child.get())) {
+        if (fgets(buf, sizeof (buf), stdout_child.get()) != nullptr) {
             output += buf;
         }
     }
-
-    //Read stderr
-    if (output.find("PASS") == std::string::npos) {
-        char buffer[256];
-        int count = read(stderr_fds[0], buffer, sizeof(buffer)-1);
-        buffer[count] = 0;
-        std::cout << buffer << std::endl;
+    while (ret == 0 && stderr_child && !feof(stderr_child.get())) {
```

Comment:
We need to explicitly check if the test passed or not. Even if the test passes, I get something like "pci id for fd x: 10ee:5005, driver (null)" in my stderr. This is what we are trying to avoid with the fds. I believe we need to check that with "output.find("PASS") == std::string::npos"

Reply:
The idea is that runShellCmd() will just run the cmd and collect and return output from both stdout and stderr back to caller. It is the caller who should decide if test has passed or not based on the content of the output. If caller decides that the test has passed, it will not show the output anyway. In this case, whatever is in stderr (now is part of combined output returned from runShellCmd()) does not matter. I tested xbutil validate on my system. When everything works, I don't see anything unexpected printed out...

Reply:
Hmm. I see... I think this should be fine then :)